kolibri/tasks/management/commands/base.py

```diff
@@ -1,45 +1,58 @@
-from tqdm import tqdm
 from collections import namedtuple
-from django.core.management.base import BaseCommand
 
+from django.core.management.base import BaseCommand
+from tqdm import tqdm
 
 Progress = namedtuple(
     'Progress',
     [
         'progress_fraction',
         'message',
         'extra_data',
+        'level',
     ]
 )
 
 
 class ProgressTracker():
 
-    def __init__(self, total=100, update_func=None):
-        self.progressbar = tqdm(total=total)
-        self.total = total
+    def __init__(self, total=100, level=0, update_callback=None):
 
+        # set default values
         self.progress = 0
+        self.message = ""
+        self.extra_data = None
 
-        # custom progress bar provided by programmer
-        if callable(update_func):
-            self.custom_update_progress_func = update_func
-        else:
-            self.custom_update_progress_func = _nullop
+        # store provided arguments
+        self.total = total
+        self.level = level
+        self.update_callback = update_callback
+
+        # initialize the tqdm progress bar
+        self.progressbar = tqdm(total=total)
 
     def update_progress(self, increment=1, message="", extra_data=None):
+
         self.progressbar.update(increment)
 
         self.progress += increment
 
-        progress_fraction = self.progress / float(self.total)
-        p = Progress(
-            progress_fraction=progress_fraction,
-            message=message,
-            extra_data=extra_data,
-        )
+        self.message = message
+
+        self.extra_data = extra_data
+
+        if callable(self.update_callback):
+            p = self.get_progress()
+            self.update_callback(p.progress_fraction, p)
```

Comment:
Why does this receive both the full `Progress` instance and a property of the same object instance? :)